### PR TITLE
Group part

### PR DESCRIPTION
**Pull request by Casmo888**
## branch name : viewall

此PR包括：
。view all running contest及該頁所有功能
。view all ended contest及該頁所有功能
。view all announce及該頁所有功能

修改先前部分：
。delete_announce加上redirect_page參數以決定該重導向detail或viewall page
。將edit announce form的寬度調整至整頁寬

### CODE REVIEW

**Comment by nickelcheng** [comment]
There are 3 "view all" buttons in ```groupDetail.html```, their only difference is url. Is there any possibility to put them into css file, so that we can eliminate redundant code?(I'm not sure, just asking)

File: group/views.py [hunk 1]
```diff
@@ -50,11 +50,10 @@
 def get_running_contest(request, group_id):
 
     group = get_group(group_id)
-
-    all_running_contest_list = []
     now = timezone.now()
 
     all_running_contest_list = group.trace_contest.filter(start_time__lte=now, end_time__gte=now)
+    all_running_contest_list = get_current_page(request, all_running_contest_list)
```
Comment (nickelcheng) [inline group/views.py]: I guess this line is for pager, right?
Can you use other variable name to avoid confusing?

File: group/views.py [hunk 2]
```diff
@@ -66,11 +65,10 @@ def get_running_contest(request, group_id):
 def get_ended_contest(request, group_id):
 
     group = get_group(group_id)
-
-    all_ended_contest_list = []
     now = timezone.now()
 
     all_ended_contest_list = group.trace_contest.filter(end_time__lte=now)
+    all_ended_contest_list = get_current_page(request, all_ended_contest_list)
```
Comment (nickelcheng) [inline group/views.py]: Ditto.

File: group/views.py [hunk 3]
```diff
@@ -83,12 +81,21 @@ def get_all_announce(request, group_id):
...
     group = get_group(group_id)
 
-    all_announce_list = group.announce.all()
+    user = validate_user(request.user)
+    user_is_owner = has_group_ownership(user, group)
+    user_is_coowner = has_group_coownership(user, group)
+    user_has_auth = user_is_owner or user_is_coowner
+
+    all_announce_list = group.announce.order_by('-id')
+    all_announce_list = get_current_page(request, all_announce_list)
```
Comment (nickelcheng) [inline group/views.py]: Ditto.

File: group/announce.py [hunk 4]
```diff
@@ -55,13 +55,16 @@ def add_announce(request, group_id):
...
 @login_required
-def delete_announce(request, announce_id, group_id):
+def delete_announce(request, announce_id, group_id, redirect_page):
     group = get_group(group_id)
 
     if has_group_ownership(request.user, group) or has_group_coownership(request.user, group):  
         get_announce(announce_id).delete()
         logger.info('Announce: User %s delete Announce %s!' % (request.user.username, announce_id))
-        return HttpResponseRedirect(reverse('group:detail', kwargs={'group_id': group_id}))
+        if redirect_page == 'detail':
```
Comment (nickelcheng) [inline group/announce.py]: I prefer that
```
if viewall:
    # redirect to view all page
else:
    # redirect to detail page
```
The difference is that my code will handle unexpected cases.
it's equal to add this line ```else: redirect to detail```

**Comment by nickelcheng** [comment]
this PR is good, @CHIAHSINHOU please test

**Comment by Casmo888** [comment]
我不太懂3個按鈕放到CSS file那個@@

File: group/views.py (hunk 1 above)
Comment (Casmo888) [inline group/views.py]: on my way

File: group/views.py (hunk 2 above)
Comment (Casmo888) [inline group/views.py]: see

File: group/views.py (hunk 3 above)
Comment (Casmo888) [inline group/views.py]: see

File: group/announce.py (hunk 4 above)
Comment (Casmo888) [inline group/announce.py]: sure, it is more better.

**Comment by Casmo888** [comment]
fixed

**Comment by CHIAHSINHOU** [comment]
This PR is good for me.

**Comment by bruce3557** [comment]
+1, going to merge

**Pull request merged**